retrievalmarket/lp2pimpl/net.go

```diff
@@ -0,0 +1,148 @@
+package lp2pimpl
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/boost/retrievalmarket"
+	"github.com/filecoin-project/boost/retrievalmarket/types"
+	"github.com/filecoin-project/go-fil-markets/shared"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	host "github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var log = logging.Logger("boost-net")
+
+// QueryV2ProtocolID is the protocol for querying information about retrieval
+// deal parameters
+const QueryV2ProtocolID = protocol.ID("/fil/retrieval/qry/2.0.0")
+
+// QueryProvider listens for incoming deal proposals over libp2p
+type QueryProvider struct {
+	ctx  context.Context
+	host host.Host
+	prov *retrievalmarket.Provider
+}
+
+const providerReadDeadline = 10 * time.Second
+const providerWriteDeadline = 10 * time.Second
+const clientReadDeadline = 10 * time.Second
+const clientWriteDeadline = 10 * time.Second
+
+// QueryClientOption is an option for configuring the libp2p storage deal client
+type QueryClientOption func(*QueryClient)
+
+// RetryParameters changes the default parameters around connection reopening
+func RetryParameters(minDuration time.Duration, maxDuration time.Duration, attempts float64, backoffFactor float64) QueryClientOption {
+	return func(c *QueryClient) {
+		c.retryStream.SetOptions(shared.RetryParameters(minDuration, maxDuration, attempts, backoffFactor))
+	}
+}
+
+// QueryClient sends retrieval queries over libp2p
+type QueryClient struct {
+	retryStream *shared.RetryStream
+}
+
+func NewQueryClient(h host.Host, options ...QueryClientOption) *QueryClient {
+	c := &QueryClient{
+		retryStream: shared.NewRetryStream(h),
+	}
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// SendQuery sends a retrieval query over a libp2p stream to the peer
+func (c *QueryClient) SendQuery(ctx context.Context, id peer.ID, query types.Query) (*types.QueryResponse, error) {
+	log.Debugw("send query", "pieceCID", query.PieceCID, "payloadCID", query.PayloadCID, "provider-peer", id)
+
+	// Create a libp2p stream to the provider
+	s, err := c.retryStream.OpenStream(ctx, id, []protocol.ID{QueryV2ProtocolID})
+	if err != nil {
+		return nil, err
+	}
+
+	defer s.Close() // nolint
+
+	// Set a deadline on writing to the stream so it doesn't hang
+	_ = s.SetWriteDeadline(time.Now().Add(clientWriteDeadline))
+	defer s.SetWriteDeadline(time.Time{}) // nolint
+
+	// Write the retrieval query to the stream
+	// Write the re to the client
+	err = types.BindnodeRegistry.TypeToWriter(query, s, dagcbor.Encode)
```

Comment:
maybe at this point it should interrogate `query` to see if it's valid, i.e. has a `PieceCID` or a `PayloadCID`, before proceeding to send it since it'll be rejected anyway